File: CODIGO_C++/ACT4_1.cpp
#include <iostream>
#include <string>

using namespace std;

//Declaracion anticipada de la clase Libro
class Libro;

//Clase Biblioteca
class Biblioteca {
public:
    //Metodo para prestar un libro
    void prestarLibro(Libro &libro);

    //Metodo para devolver un libro
    void devolverLibro(Libro &libro);
};

//Clase Libro
class Libro {
private:
    string titulo;
    bool prestado;

public:
    //Constructor de la clase Libro
    Libro(string t) : titulo(t), prestado(false) {}

    //Metodo para marcar el libro como prestado
    void marcarPrestado() {
        prestado = true;
    }

    //Metodo para marcar el libro como devuelto
    void marcarDevuelto() {
        prestado = false;
    }

    //Metodo para obtener el estado de prestamo del libro
    bool estaPrestado() {
        return prestado;
    }

    //Metodo para obtener el titulo del libro
    string obtenerTitulo() {
        return titulo;
    }
};

//Implementacion del metodo prestarLibro de la clase Biblioteca
void Biblioteca::prestarLibro(Libro &libro) {
    if (!libro.estaPrestado()) {
        libro.marcarPrestado();
        cout << "Se ha prestado el libro: " << libro.obtenerTitulo() << endl;
    } else {
        cout << "El libro " << libro.obtenerTitulo() << " ya esta prestado." << endl;
    }
}

//Implementacion del metodo devolverLibro de la clase Biblioteca
void Biblioteca::devolverLibro(Libro &libro) {
    if (libro.estaPrestado()) {
        libro.marcarDevuelto();
        cout << "Se ha devuelto el libro: " << libro.obtenerTitulo() << endl;
    } else {
        cout << "El libro " << libro.obtenerTitulo() << " no estaba prestado." << endl;
    }
}

int main() {
    //Creacion de objetos de la clase Libro
    Libro libro1("El Principito");
    Libro libro2("Don Quijote");

    //Creacion de objeto de la clase Biblioteca
    Biblioteca biblioteca;

    //Prestamo y devolucion de libros
    biblioteca.prestarLibro(libro1);
    biblioteca.prestarLibro(libro2);
    biblioteca.devolverLibro(libro1);
    biblioteca.devolverLibro(libro2);

    return 0;
}

File: CODIGO_PYTHON/ACT4_1.py
class Libro:
    def __init__(self, titulo):
        self.titulo = titulo
        self.prestado = False

    def marcar_prestado(self):
        self.prestado = True

    def marcar_devuelto(self):
        self.prestado = False

    def esta_prestado(self):
        return self.prestado

    def obtener_titulo(self):
        return self.titulo


class Biblioteca:
    def prestar_libro(self, libro):
        if not libro.esta_prestado():
            libro.marcar_prestado()
            print("Se ha prestado el libro:", libro.obtener_titulo())
        else:
            print("El libro", libro.obtener_titulo(), "ya está prestado.")

    def devolver_libro(self, libro):
        if libro.esta_prestado():
            libro.marcar_devuelto()
            print("Se ha devuelto el libro:", libro.obtener_titulo())
        else:
            print("El libro", libro.obtener_titulo(), "no estaba prestado.")


if __name__ == "__main__":
    #Creacion de objetos de la clase Libro
    libro1 = Libro("El Principito")
    libro2 = Libro("Don Quijote")

    #Creacion de objeto de la clase Biblioteca
    biblioteca = Biblioteca()

    #Prestamo y devolucion de libros
    biblioteca.prestar_libro(libro1)
    biblioteca.prestar_libro(libro2)
    biblioteca.devolver_libro(libro1)
    biblioteca.devolver_libro(libro2)
